subprojects/resources-sftp/src/main/java/org/gradle/internal/resource/transport/sftp/SftpClientFactory.java

```diff
@@ -51,10 +51,23 @@ private LockableSftpClient acquireClient(SftpHost sftpHost) {
 
     private LockableSftpClient reuseExistingOrCreateNewClient(SftpHost sftpHost) {
         List<LockableSftpClient> clientsByHost = clients.get(sftpHost);
+
+        LockableSftpClient client = null;
         if (clientsByHost.isEmpty()) {
-            return sftpClientCreator.createNewClient(sftpHost);
+            LOGGER.debug("No existing sftp clients.  Creating a new one.");
+            client = sftpClientCreator.createNewClient(sftpHost);
```

Comment:
A little, but I'm glad you bring it up because I realized that the factory is not stopping clients that haven't been explicitly released.  I've changed it so that it tracks/stops all clients instead of just idle clients.